src/delay.rs

```diff
@@ -0,0 +1,49 @@
+//! Delays
+
+use embedded_hal::timer::CountDown;
+use crate::clock::Clocks;
+use embedded_hal::blocking::delay::{DelayUs, DelayMs};
+/// Machine mode cycle counter (`mcycle`) as a delay provider
+#[derive(Copy, Clone)]
+pub struct McycleDelay {
+    core_frequency: u32
+}
+#[derive(Debug, Clone)]
+pub struct DelayError;
+
+impl McycleDelay {
+    /// Constructs the delay provider based on provided core clock frequency
+    pub fn new(freq: u32) -> Self {
+        Self {
+            core_frequency: freq
+        }
+    }
+
+    pub fn get_time() -> u64 {
```

Comment:
I don't know if get_cycle works. Or otherwise we can read using mtime::read() directly?

Reply:
get_cycle_count is used in a few other HALs. It's a much better name, thanks for prompting me to look.
@luojia65 mtime doesn't exist from what I can see, do you mean riscv::register::time?
reading from riscv::register::time triggers a hard fault (might need init?). It's also not what I want here, since the resolution is not high enough to provide microsecond delays.

Reply:
@9names In RISC-V ISA, the 'time' CSR does not really exist on hardware. This CSR takes place of CSR id, and when OS (or user application) uses it, it triggers illegal instructiom exception. The runtime (OS or SBI etc.) handles this exception, and fills in the result, as if the 'time' register is read into destination register. We are now bare-metal, and yes, this register is not used by now.
Mcycle provides a probable delay by cycle count. If we need to include hardware time, RISC-V's mtime are all memory-mapped registers. On K210 it maybe in CLINT/PLIC peripherals. In bare-metal we may directly operate on this peripheral.